#' Plot variants in genes or regions
#'
#' `plot_variants` plots the variants of an isomut file within a set of genes or
#' regions.
#'
#' @param isomut a `data.frame` or `data.table` of isomut data as generated by
#' `read_isomut`.
#' @param genes either a character vector of gene names or a `GRanges` object.
#' @param  annotation path to annotation file in `gtf` or `gff` format or `GRanges`
#' object containing the annotations.
#' @param ymax maximum of variants to plot.
#' @param fill isomut column to use for fill color (default is `mut_freq`).
#' @param filename optionally specify file name.
#' @param ... additional parameters for `ggsave` (e.g. `width` and `height`).
#'
#' @importFrom data.table as.data.table rbindlist
#' @importFrom rtracklayer import
#' @importFrom patchwork plot_layout wrap_plots wrap_elements
#' @import ggplot2
#' @importFrom grid textGrob
#' @export
plot_variants <- function(isomut, genes, annotation = gtf_scer, ymax = NULL,
                          fill = "mut_freq", filename = NULL, ...) {
  isomut <- as.data.table(isomut)
  ## load annotation
  stopifnot(is.character(annotation) || is(annotation, "GRanges"))
  if(is.character(annotation)) {
    annotation <- import(annotation)
  }
  stopifnot(is.character(genes) || is(genes, "GRanges"))
  if(is.character(genes)) {
    dfl <- lapply(genes, function(gene) {
      df <- isomut[gene_name == gene]
      df <- df[order(pos, df[[fill]])]
      mut_table <- table(df$pos)
      df$ypos <- do.call(c, lapply(mut_table, function(m) 1:m - 1))
      return(df)
    })
    names(dfl) <- genes
    grl <- lapply(genes, function(gene){
      gtf_gene <- annotation[!is.na(annotation$gene_name)]
      as.data.frame(range(gtf_gene[gtf_gene$gene_name == gene & gtf_gene$type == "gene"]))
    })
    names(grl) <- genes
  } else {
    dfl <- lapply(seq_along(genes), function(i) {
      df <- isomut[chr == as.character(seqnames(genes)[i]) & between(pos, start(genes)[i], end(genes)[i])]
      df <- df[order(pos, df[[fill]])]
      mut_table <- table(df$pos)
      if(length(mut_table)>0) df$ypos <- do.call(c, lapply(mut_table, function(m) 1:m -1))
      return(df)
    })
    grl <- lapply(seq_along(genes), function(i) {
      as.data.frame(genes[i])
    })
  }
  max_width <- max(vapply(grl, function(gr) max(gr[["width"]]), numeric(1)))
  if(is(genes, "GRanges")) genes <- seq_along(genes)
  pl <- lapply(genes, function(gene) {
    df <- dfl[[gene]]
    gene_range <- grl[[gene]]
    xmin <- gene_range$start
    xmax <- xmin + max_width
    if(is.null(ymax)) {
      mut_table <- table(df$pos)
      if(length(mut_table) > 0) {
        ymax <- max(mut_table)
      } else {
        ymax <- 1
      }
    }
    ymin <- ymax/20
    if(nrow(df) == 0) {
      p <- ggplot() +
        geom_rect(aes(xmin = start, xmax = end, ymin = -ymin, ymax = 0), color = "black", fill = "white", data = gene_range) +
        ggtitle(gene) +
        theme_classic() +
        theme(axis.title.y = element_blank()) +
        xlab(gene_range$seqnames) +
        scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (ymax + 1) * 1.1)))), limits = c(-ymin,ymax)) +
        xlim(xmin, xmax)
    } else {
      p <- ggplot() +
        geom_rect(aes(xmin = pos, xmax = pos + 5, ymin = ypos, ymax = ypos+1, fill = get(fill)), data = df) +
        geom_rect(aes(xmin = start, xmax = end, ymin = -ymin, ymax = 0), color = "black", fill = "white", data = gene_range) +
        ggtitle(gene) +
        theme_classic() +
        theme(axis.title.y = element_blank()) +
        xlab(gene_range$seqnames) +
        scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (ymax + 1) * 1.1)))), limits = c(-ymin,ymax)) +
        xlim(xmin, xmax)
    }
    if(is.factor(isomut[[fill]])) {
      p <- p +
        scale_fill_discrete(drop = FALSE) +
        guides(fill=guide_legend(title=fill))
    } else {
      if(fill=="mut_freq") {
        fillmin <- 0
        fillmax <- 1
      } else {
        fillmin <- min(map_dbl(dfl, ~min(.[[fill]])))
        fillmax <- max(map_dbl(dfl, ~max(.[[fill]])))
      }
      p <- p +
        scale_fill_continuous(low = "yellow", high = "red", limits = c(fillmin, fillmax)) +
        guides(fill=guide_colorbar(title=fill))
    }
    return(p)
  })
  p <- wrap_elements(textGrob('number of variants', rot = 90)) + wrap_plots(pl, ncol = 1)
  p <- p + plot_layout(widths = c(1,20), guides = "collect") & theme(legend.position = 'right')
  if(!is.null(filename)) ggsave(filename, p, ...)
  return(p)
}
