#' Plot variants in genes or regions
#'
#' `plot_variants` plots the variants of an isomut file within a set of genes or
#' regions.
#'
#' @param isomut a `data.frame` or `data.table` of isomut data as generated by
#' `read_isomut`.
#' @param regions either a character vector of gene names or a `GRanges` object.
#' @param  annotation path to annotation file in `gtf` or `gff` format or `GRanges`
#' object containing the annotations.
#' @param ymax maximum of variants to plot.
#' @param fill isomut column to use for fill color (default is `mut_freq`).
#' @param bar_width width of the bars (default 5).
#' @param flank additional space left and right of genes (in bp).
#' @param showGenes whether to show genes in the plot (only works with annotation).
#' @param fontsize font size for gene and chromosome names.
#' @param dy width of gene/chromosome bars.
#' @param filename optionally specify file name.
#' @param ... additional parameters for `ggsave` (e.g. `width` and `height`).
#'
#' @importFrom data.table as.data.table rbindlist
#' @importFrom rtracklayer import
#' @importFrom patchwork plot_layout wrap_plots wrap_elements
#' @importFrom GenomicRanges seqnames GRangesList
#' @import ggplot2
#' @importFrom grid textGrob
#' @export
plot_variants <- function(isomut, regions, annotation = gtf_scer, ymax = NULL,
                          fill = NULL, bar_width = NULL, flank = 50,
                          showGenes = TRUE, fontsize = NULL, dy = NULL,
                          filename = NULL,
                          ...) {
  if(is.null(fontsize)) fontsize <- 8/length(regions)
  isomut <- as.data.table(isomut)
  ## load and prepare annotation
  stopifnot(is.character(annotation) || is(annotation, "GRanges"))
  if(is.character(annotation)) {
    annotation <- import(annotation)
  }
  gtf_gene <- annotation[annotation$type == "gene"]
  gtf_gene$gene_name[is.na(gtf_gene$gene_name)] <- gtf_gene$gene_id[is.na(gtf_gene$gene_name)]
  names(gtf_gene) <- gtf_gene$gene_name
  ## prepare regions
  stopifnot(is.character(regions) || is(regions, "GRanges"))
  if(is.character(regions)) {
    regions <- gtf_gene[regions]
    strand(regions) <- "*"
    regions <- resize(regions, width = width(regions) + 2 * flank, fix = "center")
  }
  grl <- split(regions, factor(regions, levels = as.factor(regions)))
  names(grl) <- names(regions)
  grl_genes <- GRangesList(lapply(grl, function(gr) gtf_gene[gtf_gene %over% gr]))

  max_width <- max(vapply(grl, function(gr) max(width(gr)), numeric(1)))
  if(is.null(bar_width)) bar_width <- max_width/500
  ## count variants
  dfl <- lapply(grl, function(gr) {
    df <- isomut[chr == as.character(seqnames(gr)) & between(pos, start(gr), end(gr))]
    if(is.null(fill)) {
      df <- df[order(pos)]
    } else {
      df <- df[order(pos, df[[fill]])]
    }
    mut_table <- table(df$pos)
    if(nrow(df) > 0) df$ypos <- do.call(c, lapply(mut_table, function(m) 1:m - 1))
    return(df)
  })
  ## plot ranges
  pl <- lapply(names(grl), function(name) {
    gr <- as.data.frame(grl[[name]])
    df <- dfl[[name]]
    df_genes <- as.data.frame(grl_genes[[name]])
    xmin <- gr$start
    xmax <- xmin + max_width
    df_genes$start <- pmax(df_genes$start, xmin)
    df_genes$end <- pmin(df_genes$end, xmax)
    df_genes$width <- abs(df_genes$end - df_genes$start)
    if(is.null(ymax)) {
      if(nrow(df)>0) {
        ymax <- max(df$ypos) + 1
      } else {
        ymax <- 1
      }
    }
    if(is.null(dy)) dy <- ymax/10
    p <- ggplot() +
      ggtitle(name) +
      theme_classic() +
      theme(axis.title.y = element_blank()) +
      xlab("bp")
    if(nrow(df)>0) {
      if(is.null(fill)) {
        p <- p +
          geom_rect(aes(xmin = pos, xmax = pos + bar_width, ymin = ypos, ymax = ypos+1), data = df)
      } else {
        p <- p +
          geom_rect(aes(xmin = pos, xmax = pos + bar_width, ymin = ypos, ymax = ypos+1, fill = get(fill)), data = df)
      }
    }
    p <- p + scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (ymax + 1) * 1.1)))), limits = c(-2*dy, ymax)) +
      geom_rect(aes(xmin = xmin, xmax = xmax, ymin = -dy, ymax = 0),
                fill = "black", alpha = 0.2) +
      geom_text(aes(x = start, y = -dy/2, label = seqnames), data = gr,
                size = fontsize, hjust = 0, nudge_x = max_width/100,
                fontface = "bold", ) +
      scale_x_continuous(expand = c(0,0))
    if(nrow(df_genes)>0 && showGenes) {
      p <- p +
        geom_rect(aes(xmin = start, xmax = end, ymin = -2*dy, ymax = -dy),
                  fill = "blue", alpha = 0.3, data = df_genes) +
        geom_text(aes(x = start, y = -dy*1.5, label = gene_name), data = df_genes,
                  size = fontsize, hjust = 0, nudge_x = max_width/100,
                  fontface = "bold")
    }
    if(!is.null(fill)) {
      if(is.factor(isomut[[fill]])) {
        p <- p +
          scale_fill_discrete(drop = FALSE) +
          guides(fill=guide_legend(title=fill))
      } else {
        if(fill=="mut_freq") {
          fillmin <- 0
          fillmax <- 1
        } else {
          fillmax <- max(vapply(dfl, function(df) max(df[[fill]]), numeric(1)))
          fillmin <- min(vapply(dfl, function(df) min(df[[fill]]), numeric(1)))
        }
        p <- p +
          scale_fill_continuous(low = "yellow", high = "red", limits = c(fillmin, fillmax)) +
          guides(fill=guide_colorbar(title=fill))
      }
    }
    return(p)
  })
  p <- wrap_elements(textGrob('number of samples', rot = 90)) + wrap_plots(pl, ncol = 1)
  p <- p + plot_layout(widths = c(1,20), guides = "collect") & theme(legend.position = 'right')
  if(!is.null(filename)) ggsave(filename, p, ...)
  return(p)
}
